Add more tap/target maintenance content (#1133)
meltano/hub: static/markdown/tap-target-maintenance.md

This approach may solve the short term problem but requires you to take on the burden of maintaining the repo moving forward.
By contributing back to the upstream you benefit the community and you can free yourself from maintaining your variant moving forward.
If everyone in our community puts their effort into maintaining the same connectors then we all benefit from less work and better quality.

### Favor Generically Useful vs Custom or Hardcoded Logic

Sometimes it's tempting to build a connector with only your exact needs baked in. For example only adding the streams you need or hard coding organization specific logic like account names, how credentials are sourced, etc. But with a little extra effort you can make it generically useful enough to attract other community members to use and help maintain your code.

- Add as much stream coverage as you can
- All configs should flow through the config.json input file
- No hardcoded values and logic, or at least not by default

### Document Settings Well

The most challenging part of using a community maintained connector is configuring it properly.
Always make sure to document your configurations well.

> SDK based connectors have a [--about command option](https://sdk.meltano.com/en/latest/implementation/cli.html#about) that prints out the documented settings and descriptions from `tap.py`. MeltanoHub scrapes these settings and descriptions for building documentation so you should make sure to keep them up to date.

Reviewer: ```suggestion
SDK based connectors have a [--about command option](https://sdk.meltano.com/en/latest/implementation/cli.html#about) that prints out the documented settings and descriptions from `tap.py`. 
Meltano Hub uses these settings and descriptions for documentation, so ensuring they're up to date makes it more useful for everyone.
```

For non-SDK based connectors you will commonly find an example config.json in the repo.
The Meltano community has also invested in documenting these on MeltanoHub for your convenience.

### Advertise It On Meltano Hub

> Once your connector is functioning and ready to go, make sure you get it listed on MeltanoHub (see [docs](/add-a-tap)) so others can find it.

Reviewer: ```suggestion
Once your connector is functioning and ready to go, make sure you get it listed on Meltano Hub (see [docs](/add-a-tap)) so others can find it.
```

More users means more contributors and maintainers for your code base!

> ### Use Meltano SDK

Reviewer: ```suggestion
### Use the Meltano Singer SDK
```

> We might be a little biased on this one but using an SDK allows you to reduce code, increase reliability, comply with the Singer Spec, gain features with an SDK version upgrade, and work well within the Meltano community.

Reviewer: ```suggestion
Using the SDK allows you to reduce code, increase reliability, comply with the Singer Spec, gain features with an SDK version upgrade, and work well within the Meltano community. The SDK is maintained by Meltano and receives regular updates. Upgrading a tap built on the SDK is much easier than one that isn't.
```

## Common Patterns

> Since Singer is only a spec (see Meltano's [Singer Spec docs](http://localhost:8080/singer/docs#singer-spec)) for how data should be formatted and passed between connectors there have become many common patterns, tools, and SDKs to help speed the process up.

Reviewer: ```suggestion
Since Singer is a primarily a specification for data formatting and transfer(see Meltano's [Singer Spec docs](http://localhost:8080/singer/docs#singer-spec)) there have become many common patterns, tools, and SDKs to help build connectors and pipelines more efficiently.
```

#### [singer-io](https://github.com/singer-io)

- Mostly from scratch implementations, using the [singer-python](https://github.com/singer-io/singer-python) library for some shared methods
- Commonly have most of the code in the module's `__init__.py` file with a `main` method

> - Logic for parsing cli inputs uses argparse and is required by every connector

Reviewer: ```suggestion
- Logic for parsing CLI inputs uses argparse and is required by every connector
```

- A `REQUIRED_CONFIG_KEYS` list to validate required config values
- Dependencies are usually in the setup.py vs in requirements.txt
- Very few target examples because they we're originally built to always send data to the [Stitch](https://www.stitchdata.com/) service with [target-stitch](https://github.com/singer-io/target-stitch)
- Usually unit tests are limited
- The developer needs to know internal rules like when its safe to write state messages or how to handle ACTIVATE_VERSION messages
- The `catalog` input file was originally called `properties` so some will accept either
- [Getting started guide](https://github.com/singer-io/getting-started)
- [tap template](https://github.com/singer-io/singer-tap-template)

> #### [Meltano SDK](https://sdk.meltano.com/en/latest/) Based Connectors

Reviewer: ```suggestion
#### [Meltano Singer SDK](https://sdk.meltano.com/en/latest/) Based Connectors
```

Let's go ahead and move this section to the top.

> A software development kit that abstracts many of the challenges of building Singer connectors so the developer doesn't need to know about them and also provides accelerators to speed up the process of writing a new connector. See the [Meltano SDK](https://sdk.meltano.com/en/latest/) docs for more details. Example of some features:

Reviewer: ```suggestion
A software development kit that abstracts many of the challenges of building Singer connectors so the developer doesn't need to know about them and also provides accelerators to speed up the process of writing a new connector. 
See the [Meltano SDK](https://sdk.meltano.com/en/latest/) docs for more details. Example of some features:
```

- cookiecutter templates to seed a connector repo
- manages translating records into Singer spec compliant messages
- methods for common API functionality (i.e. parent child relationships, pagination, retry, post processing, etc.)

> - methods for common SQL functionality

Reviewer: ```suggestion
- methods for common SQL functionality
- native support for advanced features such as BATCH messaging and the ACTIVATE_VERSION sync strategy
```

## Evaluating a tap or target

These are things you can look for when evaluating a tap or target.
Many of these recommendations can be used for evaluating open source projects in general.

You should start by looking on MeltanoHub to find and evaluate connectors.
Although a connector can still be listed if its not well maintained so click through to the repo to inspect it closer if you're unsure.
None of these attributes alone will tell you what the best connectors are, so look at many of them to make a holistic evaluation.

- <u>Organization backed connectors</u>: If its in the namespace of `meltano`/`meltanolabs`/`transferwise`/`singer-io`/ etc. its usually a good bet that the connector is maintained because one of the organizations is invested in it's success. Although this isn't always the case, especially with some of the older singer-io variants.
- <u>MeltanoHub default</u>: The Meltano community does it's best to label the most active variant as the default on MeltanoHub. Although it's still possible that the best option is unmaintained.

> - <u>Meltano SDK</u>: If it uses the Meltano SDK then its a good sign that its relatively recent and has good support for common features. If there are many TODOs still in the repo that came with the cookiecutter it might indicate that the connector was never finished or is a hobby project.

Reviewer: ```suggestion
- <u>Meltano Singer SDK</u>: If it uses the Meltano Singer SDK then its a good sign that its relatively recent and has good support for common features. If there are many TODOs still in the repo that came with the cookiecutter it might indicate that the connector was never finished or is a hobby project.
```

- <u>Stream coverage</u>: You can evaluate 2 variants against each other based on stream coverage. Sometimes a new variant is written that has more features but is only supporting a small number of streams.

> - <u>Latest commit timestamp</u>: If there are commits in the last couple months then its probably decently maintained.

Reviewer: ```suggestion
- <u>Latest commit timestamp</u>: If there are commits in the last couple months then its probably decently maintained. However, many APIs can be quite stable, so a lack of commits does not mean the connector does not work.
```

- Add as much stream coverage as you can
- All configs should flow through the config.json input file
- No hardcoded values and logic, or at least not by default

### Document Settings Well

The most challenging part of using a community maintained connector is configuring it properly.
Always make sure to document your configurations well.

SDK based connectors have a [--about command option](https://sdk.meltano.com/en/latest/implementation/cli.html#about) that prints out the documented settings and descriptions from `tap.py`.
Meltano Hub uses these settings and descriptions for documentation, so ensuring they're up to date makes it more useful for everyone.

For non-SDK based connectors you will commonly find an example config.json in the repo.
The Meltano community has also invested in documenting these on MeltanoHub for your convenience.

### Advertise It On Meltano Hub

Once your connector is functioning and ready to go, make sure you get it listed on Meltano Hub (see [docs](/add-a-tap)) so others can find it.
More users means more contributors and maintainers for your code base!

### Use the Meltano Singer SDK

> Using the SDK allows you to reduce code, increase reliability, comply with the Singer Spec, gain features with an SDK version upgrade, and work well within the Meltano community. The SDK is maintained by Meltano and receives regular updates. Upgrading a tap built on the SDK is much easier than one that isn't.

Reviewer 2: I wanted to add something specifically about the benefits of the new testing framework, and I tried to match the tone of the paragraph. Feel free to wordsmith as you see fit.

```suggestion
Using the SDK allows you to reduce code, increase reliability, comply with the Singer Spec, gain features with an SDK version upgrade, and work well within the Meltano community. The SDK is maintained by Meltano and receives regular updates. The SDK also has tons of [valuable pre-written tests](https://sdk.meltano.com/en/latest/testing.html) that you can take advantage of with only a few lines of code, allowing you to maintain and upgrade your tap or target with more confidence and less manual testing effort.
```
